haskell/src/Parser.hs

```diff
@@ -0,0 +1,43 @@
+module Parser (evaluateExpression) where
+
+import Text.ParserCombinators.Parsec
+import Nodes
+
+evaluateExpression  :: String -> Maybe ([Node])
+evaluateExpression input =
+    case parse parser "(unknown)" input of
+    Left parseError -> Nothing
+    Right result -> Just result
+
+parser :: Parser [Node]
+parser = expressionParser
+
+expressionParser :: Parser [Node]
+expressionParser = do
+    number <- numberParser
+    spaces
+    nodes <- many furtherExpressionParser
+    return $ [NumberNode (read number)] ++ (concat nodes)
+
+furtherExpressionParser :: Parser [Node]
+furtherExpressionParser = do
```

Comment:
If you can define `operatorParser :: Parser Node` and `numberParser :: Node`, then you can define this more easily:

``` haskell
furtherExpressionParser :: Parser [Node]
furtherExpressionParser = sequence [numberParser, operatorParser]
```

Sequence will go from `[Parser Node]` to `Parser [Node]`.